File: main/client/src/lock/Lock.scala
package mill.main.client.lock

abstract class Lock extends AutoCloseable {

  @throws[Exception]
  def lock(): Locked

  @throws[Exception]
  def tryLock(): TryLocked

  @throws[Exception]
  def await(): Unit = {
    lock().release()
  }

  /**
   * Returns `true` if the lock is *available for taking*
   */
  @throws[Exception]
  def probe(): Boolean

  @throws[Exception]
  def delete(): Unit = {}

}

object Lock {
  @throws[Exception]
  def file(path: String): Lock = new FileLock(path)

  def memory(): Lock = new MemoryLock()
}

File: runner/server/client/src/mill/client/lock/Lock.java
package mill.client.lock;

public abstract class Lock implements AutoCloseable {

  public abstract Locked lock() throws Exception;

  public abstract TryLocked tryLock() throws Exception;

  public void await() throws Exception {
    lock().release();
  }

  /**
   * Returns `true` if the lock is *available for taking*
   */
  public abstract boolean probe() throws Exception;

  public void delete() throws Exception {}

  public static Lock file(String path) throws Exception {
    return new FileLock(path);
  }

  public static Lock memory() {
    return new MemoryLock();
  }
}
